### PR TITLE
Adding methods to locate folders and files in VS installs

### PR DESCRIPTION
**Pull request by olgaark**
VC build needs to be able to use toolsets installed by older VS versions (native multitargeting). Before VS2017 the vc targets locations were written in the registry. To figure out Vs2017 install locations we have to call the installer API.  There also might be several instances on the machine and we need to be able to use a toolset installed in any of the instances.

Here is how the added functions are going to be used in cpp props:
 
```xml
<PropertyGroup>
    <_VCTargetPath15Folders>$([Microsoft.Build.Utilities.ToolLocationHelper]::GetFoldersInVSInstallsAsString('15.0', '16.0', 'Common7\IDE\VC\VCTargets'))</_VCTargetPath15Folders>
    <_VCTargetsPathForToolset Condition="'$(_VCTargetPath15Folders)' != ''">$([Microsoft.Build.Utilities.ToolLocationHelper]::FindRootFolderWhereAllFilesExist($(_VCTargetPath15Folders), $(_RelativeToolsetFiles)))</_VCTargetsPathForToolset>
  </PropertyGroup>
```

### CODE REVIEW

File: src/Utilities/ToolLocationHelper.cs [hunk 1]
```diff
@@ -1381,6 +1386,133 @@ public static string GetLatestSDKTargetPlatformVersion(string sdkIdentifier, str
...
+                foreach (var root in roots)
+                {
+                    allFilesFound = true;
+                    foreach (var file in files)
+                    {
+                        try
+                        {
+                            string fullPath = Path.Combine(root, file);
+
+                            if (!FileSystems.Default.FileExists(fullPath))
```
Comment (rainersigwald) [inline src/Utilities/ToolLocationHelper.cs]: This seems like a place where we'd want to use a cached evaluation context. But I don't know if it's plumbed into the right places.

(This can wait; it's an optimization)

File: ref/Microsoft.Build.Utilities.Core/netstandard/Microsoft.Build.Utilities.Core.cs [hunk 2]
```diff
@@ -282,6 +283,8 @@ public static partial class ToolLocationHelper
         public static string GetDotNetFrameworkSdkRootRegistryKey(Microsoft.Build.Utilities.TargetDotNetFrameworkVersion version) { throw null; }
         public static string GetDotNetFrameworkSdkRootRegistryKey(Microsoft.Build.Utilities.TargetDotNetFrameworkVersion version, Microsoft.Build.Utilities.VisualStudioVersion visualStudioVersion) { throw null; }
         public static string GetDotNetFrameworkVersionFolderPrefix(Microsoft.Build.Utilities.TargetDotNetFrameworkVersion version) { throw null; }
+        public static System.Collections.Generic.IEnumerable<string> GetFoldersInVSInstalls(System.Version minVersion=null, System.Version maxVersion=null, string subFolder=null) { throw null; }
+        public static string GetFoldersInVSInstallsAsString(string minVersionString=null, string maxVersionString=null, string subFolder=null) { throw null; }
```
[the remark this answers is not in the document]
Comment (olgaark) [reply ref/Microsoft.Build.Utilities.Core/netstandard/Microsoft.Build.Utilities.Core.cs]: The couple of files under refs were automatically updated (by the solution build, I guess). Tell me if I need to revert them.

[the remark this answers is not in the document]
Comment (rainersigwald) [reply ref/Microsoft.Build.Utilities.Core/netstandard/Microsoft.Build.Utilities.Core.cs]: No, they're a correct result of the changes you're making. I'm asking if we should expose this functionality in Core, or `#if` the implementation to make it full-framework only.

[the remark this answers is not in the document]
Comment (olgaark) [reply ref/Microsoft.Build.Utilities.Core/netstandard/Microsoft.Build.Utilities.Core.cs]: Do you want me to add #if FEATURE_VISUALSTUDIOSETUP to hide the methods in Core? Or something else?

[the remark this answers is not in the document]
Comment (rainersigwald) [reply ref/Microsoft.Build.Utilities.Core/netstandard/Microsoft.Build.Utilities.Core.cs]: I think it's either that, or `#if FEATURE_VISUALSTUDIOSETUP` _inside_ the methods and have them just return "nothing found" in the Core case. Thoughts @Microsoft/msbuild-maintainers?

File: src/Utilities/ToolLocationHelper.cs [hunk 3]
```diff
@@ -1381,6 +1386,133 @@ public static string GetLatestSDKTargetPlatformVersion(string sdkIdentifier, str
...
+        /// <param name="minVersion">Optional. If specified, only VS instances with this or bigger version will be included.</param>
+        /// <param name="maxVersion">Optional. If specified, only VS instances with smaller versions will be included. For instance, 16.0 means that only 15.* versions will be included.</param>
+        /// <param name="subFolder">Optional. If specified, the returned list will contain [VSInstallDir][subFolder] paths</param>
+        /// <returns>Returns folders in VS installs in VS version descending order</returns>
+        public static IEnumerable<string> GetFoldersInVSInstalls(Version minVersion = null, Version maxVersion = null, string subFolder = null)
+        {
+            string versionRange = $"{minVersion?.ToString() ?? string.Empty};{maxVersion?.ToString() ?? string.Empty}";
+            string[] vsInstallFolders;
+
+            lock (s_locker)
```
Comment (cdmihai) [inline src/Utilities/ToolLocationHelper.cs]: The locking pattern looks very amenable to ConcurrentDictionary.GetOrAdd

**Comment by cdmihai** [comment]
Would it make better sense to merge these two APIs into a single one like `ResolveFilesAgainstVSInstances` which gets a set of relative files and returns the set normalized against the first VS instance which contains all of them?

**Comment by olgaark** [comment]
I am planning to use FindRootFolderWhereAllFilesExist for 14 and older VS versions  as well as additional paths specified for a project (new feature in Dev16), so I'd like to have it separately from VsInstalls method.

File: src/Utilities/ToolLocationHelper.cs (hunk 3 above)
Comment (olgaark) [reply src/Utilities/ToolLocationHelper.cs]: Is it required? The other methods in this file use locker.

**Comment by olgaark** [comment]
I thought about unittests, but I could not figure out how I actually do anything meaningful there. Can you help?

**Comment by olgaark** [comment]
Or are you saying to add a test for Unix and make sure an empty list is returned?

File: src/Utilities/ToolLocationHelper.cs (hunk 3 above)
Comment (cdmihai) [reply src/Utilities/ToolLocationHelper.cs]: I think it would be something to keep in mind if we start seeing contention on the lock, since the lock protects unrelated pieces of data.
